### PR TITLE
TreehouseTester

### PR DESCRIPTION
**Pull request by swankjesse**
This is a new testing facet that launches real code with a real zipline.

In a follow-up PR I intend to put this to work to confirm that we don't hold references to various types beyond the required lifetime.

### CODE REVIEW

File: redwood-treehouse-host/src/appsTest/kotlin/app/cash/redwood/treehouse/TreehouseTesterTest.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    content.bind(view)
+
+    content.awaitContent(1)
+    val textInputValue = view.views.single() as TextInputValue
+    assertThat(textInputValue.text).isEqualTo("what would you like to see?")
+    textInputValue.onChange!!.invoke("TreehouseTesterTestHappyPathStep2")
+
+    tester.sendFrame()
+    content.awaitContent(2)
```
Comment (JakeWharton) [inline redwood-treehouse-host/src/appsTest/kotlin/app/cash/redwood/treehouse/TreehouseTesterTest.kt:51]: Is this waiting one change or two? I can't tell if the count is kept and so you have to increment it each time, or if it's reset and you could use 1 over and over if you wanted.

File: redwood-treehouse-host/build.gradle [hunk 2]
```diff
@@ -35,10 +35,21 @@ kotlin {
...
+        implementation projects.redwoodLayoutTesting
+        implementation projects.redwoodLazylayoutTesting
+        implementation projects.testApp.schema.protocolHost
+        implementation projects.testApp.schema.testing
+      }
+      if (!rootProject.hasProperty('redwoodNoApps')) {
+        kotlin.srcDir('src/appsTest/kotlin')
+        dependencies {
+          implementation projects.testApp.presenterTreehouse
+        }
```
Comment (JakeWharton) [inline redwood-treehouse-host/build.gradle:52]: I'm pretty sure this means this code isn't going to be tested on CI. You would need to add `:redwood-treehouse-host:check` to the 'test-app' shard.

File: redwood-treehouse-host/build.gradle [hunk 3]
```diff
@@ -80,3 +95,10 @@ android {
     androidResources = true
   }
 }
+
+// TreehouseTester expects to load our test application.
+if (!rootProject.hasProperty('redwoodNoApps')) {
+  tasks.withType(Test.class).configureEach {
+    it.dependsOn(":test-app:presenter-treehouse:compileDevelopmentExecutableKotlinJsZipline")
```
Comment (JakeWharton) [inline redwood-treehouse-host/build.gradle:102]: 😬

**Review by JakeWharton — APPROVED** [review]

File: redwood-treehouse-host/build.gradle (hunk 2 above)
Comment (swankjesse) [reply redwood-treehouse-host/build.gradle:52]: Ooooh good catch

Comment (swankjesse) [reply redwood-treehouse-host/build.gradle:52]: On my follow-up list.

File: redwood-treehouse-host/src/appsTest/kotlin/app/cash/redwood/treehouse/TreehouseTesterTest.kt (hunk 1 above)
Comment (swankjesse) [reply redwood-treehouse-host/src/appsTest/kotlin/app/cash/redwood/treehouse/TreehouseTesterTest.kt:51]: It’s waiting for version 2 of the content.

Comment (swankjesse) [reply redwood-treehouse-host/src/appsTest/kotlin/app/cash/redwood/treehouse/TreehouseTesterTest.kt:51]: (I’ll find a way to make this API more clear.)

File: redwood-treehouse-host/build.gradle (hunk 2 above)
Comment (swankjesse) [reply redwood-treehouse-host/build.gradle:52]: https://github.com/cashapp/redwood/pull/2065

**Pull request merged**